src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/DocumentBasedFixAllProvider.cs

```diff
@@ -106,9 +115,26 @@ private static async Task<ImmutableArray<Diagnostic>> DetermineDiagnosticsAsync(
         {
             using var _ = progressTracker.ItemCompletedScope();
 
-            return fixAllContext.Document != null
-                ? await fixAllContext.GetDocumentDiagnosticsAsync(fixAllContext.Document).ConfigureAwait(false)
-                : await fixAllContext.GetAllDiagnosticsAsync(fixAllContext.Project).ConfigureAwait(false);
+            if (fixAllContext.Document != null)
+            {
+                if (fixAllContext.State.FixAllSpans.IsEmpty)
+                {
+                    return await fixAllContext.GetDocumentDiagnosticsAsync(fixAllContext.Document).ConfigureAwait(false);
+                }
+                else
+                {
+                    var diagnostics = ImmutableArray<Diagnostic>.Empty;
```

Comment:
Use a builder?

Reply:
So, for 99% cases we expect a single FixAll span entry here. We get multiple FixAll spans within a document only for the case where a single named type has multiple partial definitions defined in the same file, which is almost never the case in real code bases. I don't think we need to optimize for this corner case.

Reply:
my general concern is simply that the builder approach is just as easy to write, and works well for the 1 to many case.  I agree it's unlikely, but really rthe ImmutableArray.AddRange approach really stands out like a sort thumb as "this is just waiting for an unfortunate case.  Note: sam's TemporaryArray is great here.  just do:

```c#
using var diagnostics = TemporaryArray<Diagnostic>Empty;
```

As fast for the normal case (<4 items), and no pathlogical dropoff.